mne/viz/evoked.py

```diff
@@ -1698,258 +1636,215 @@ def _evoked_sensor_legend(info, picks, ymin, ymax, show_sensors, ax):
+        orig_bbox = ax.get_position()
         divider = make_axes_locatable(ax)
-        ax_cb = divider.append_axes("right", size="5%", pad=0.05)
-        if colors_are_float:
-            ax_cb.imshow(the_colors[:, np.newaxis, :], interpolation='none',
-                         aspect=.05)
-            color_ticks = np.array(list(set(colors.values()))) * 100
-            ax_cb.set_yticks(color_ticks)
-            ax_cb.set_yticklabels(color_ticks)
-        else:
-            ax_cb.imshow(the_colors[:, np.newaxis, :], interpolation='none')
-            ax_cb.set_yticks(np.arange(len(the_colors)))
-            ax_cb.set_yticklabels(np.array(color_conds)[color_order])
-        ax_cb.yaxis.tick_right()
-        ax_cb.set(xticks=(), ylabel=cmap_label)
-
+        cax = divider.append_axes('right', size='5%', pad=0.1)
+        cb = ColorbarBase(cax, cmap=cmap, norm=None,
+                          orientation='vertical')
+        cb.set_label(cmap_label)
+        ticks = list()
+        ticklabels = list()
+        for cond, tick_loc in legend_tick_locs.items():
+            # handle conditions with same color/location
+            if tick_loc in ticks:
+                idx = ticks.index(tick_loc)
+                if ticklabels[idx] != cond:
+                    ticklabels[idx] = '\n'.join([ticklabels[idx], cond])
+            else:
+                ticks.append(tick_loc)
+                ticklabels.append(cond)
+        order = np.argsort(ticks)
+        cb.set_ticks(np.array(ticks)[order])
+        cb.set_ticklabels(np.array(ticklabels)[order])
+        cax.yaxis.tick_right()
+        # shrink colorbar if discrete colors
+        color_vals = list(colors.values())
+        if all([isinstance(_color, Integral) for _color in color_vals]):
+            fig = ax.get_figure()
+            fig.canvas.draw()
+            fig_aspect = np.divide(*fig.get_size_inches())
+            plot_bbox = ax.get_position()
+            cax_width = 0.75 * (orig_bbox.xmax - plot_bbox.xmax)
+            # add extra space for multiline colorbar labels
+            h_mult = max(2, max([len(lab.split('\n')) for lab in ticklabels]))
+            cax_height = len(set(color_vals)) * h_mult * cax_width / fig_aspect
+            x0 = orig_bbox.xmax - cax_width
+            y0 = (plot_bbox.ymax + plot_bbox.ymin - cax_height) / 2
+            x1 = orig_bbox.xmax
+            y1 = y0 + cax_height
+            new_bbox = Bbox([[x0, y0], [x1, y1]])
+            cax.set_axes_locator(None)
+            cax.set_position(new_bbox)
+    # legend params
+    ncol = 1 + (len(lines) // (4 if split_legend else 5))
+    legend_params = dict(loc=loc, frameon=True, ncol=ncol)
+    if do_topo and isinstance(legend, bool):
+        legend_params.update(loc='lower right', bbox_to_anchor=(1, 1))
+    if draw_legend:
+        labels = [li.get_label() for li in lines]
+        ax.legend(lines, labels, **legend_params)
+    # we needed the lines for the legends, but now we can kill them
```

Comment:
I think with recent matplotlib, there is an easier way of doing this, without drawing fake lines.

Reply:
I'll look into that.

Reply:
Drawing fake lines is still the recommended method for custom legends. See https://matplotlib.org/3.1.0/tutorials/intermediate/legend_guide.html#proxy-legend-handles

Reply:
Hm, ok.